### PR TITLE
[hosting-logs] Add WithLogging

### PR DESCRIPTION
**Pull request by CodeBlanch**
Relates to #4433

## Changes

* Adds `WithLogging` extension on `OpenTelemetryBuilder`

## Details

We're getting into the more interesting bits of this effort!

The current design on this PR is these 3 things are equivalent:

* Traditional API:

   ```csharp
   appBuilder.Logging.AddOpenTelemetry();
   ```

* New API:

   ```csharp
   appBuilder.Services.AddOpenTelemetry().WithLogging();
   ```

* Both APIs:

   ```csharp
   // Both may be called. Only a single LoggerProvider (OTel) and ILoggerProvider (.NET) will exist
   appBuilder.Logging.AddOpenTelemetry();
   appBuilder.Services.AddOpenTelemetry().WithLogging();
   ```

`WithLogging` doesn't necessarily need to exist and/or doesn't need to turn on the ILogger integration. Discussed this with @noahfalk a bit and we felt this was reasonable. But open to ideas/discussions/suggestions.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated

### CODE REVIEW

**Review by alanwest — APPROVED** [review]
> WithLogging doesn't necessarily need to exist and/or doesn't need to turn on the ILogger integration

I understand what you're saying, but I will assert that it does need to exist. I think consistency is important. OpenTelemetry has conditioned folks to think about telemetry as metrics, traces, and logs. Good to reenforce that conditioning 😄

**Review by reyang — APPROVED** [review]

**Comment by CodeBlanch** [comment]
FYI I pushed an update so that `OpenTelemetryBuilder.ConfigureResource` will also apply to logging.

On that topic...

What makes `OpenTelemetryBuilder` exciting is it gives us a place to do cross-cutting things. @noahfalk and I discussed possibly in the future having...

```csharp
appBuilder.Services.AddOpenTelemetry()
  .ConfigureResource(...)
  .AddOtlpExporter() // Turns on OtlpExporter for all signals
  .WithTracing()
  .WithMetrics()
  .WithLogging();
```

File: test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetryServicesExtensionsTests.cs [hunk 1]
```diff
@@ -326,6 +327,129 @@ public void AddOpenTelemetry_WithMetrics_NestedResolutionUsingConfigureTest()
...
+
+        services.AddOpenTelemetry().WithLogging(builder => { });
+
+        services.AddOpenTelemetry().WithLogging(builder => { });
+
+        using var serviceProvider = services.BuildServiceProvider();
+
+        Assert.NotNull(serviceProvider);
+
+        var tracerProviders = serviceProvider.GetServices<LoggerProvider>();
```
Comment (utpilla) [inline test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetryServicesExtensionsTests.cs:343]: nit: 
```suggestion
        var =loggerProviders = serviceProvider.GetServices<LoggerProvider>();
```

Comment (CodeBlanch) [reply test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetryServicesExtensionsTests.cs:343]: Thanks I will fix. Also noticed we don't have tests for `OpenTelemetryBuilder.ConfigureResource` I'll add one/some.

File: test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetryBuilderTests.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+{
+    [Fact]
+    public void ConfigureResourceTest()
+    {
+        var services = new ServiceCollection();
+
+        services
+            .AddOpenTelemetry()
+            .ConfigureResource(r => r.AddResource(new Resource(new Dictionary<string, object> { ["key1"] = "value1" })))
+            .WithLogging()
```
Comment (utpilla) [inline test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetryBuilderTests.cs:36]: It's probably a good idea to also add the signal specific `ConfigureResource` calls to test that they continue to work correctly when used with `OpenTelemetryBuilder.ConfigureResource`.

**Pull request merged**